File: apps/api/src/connections/connections.service.ts
import { HttpStatus, Injectable } from '@nestjs/common';
import { isUUID } from 'class-validator';
import { Device, DeviceConnection } from '@prisma/client';
import { DeviceConnectionDto, PaginatedResponse, WS_EVENTS } from '@nodescope/shared';
import { AuditService } from '../audit/audit.service';
import { NodeScopeException } from '../common/filters/global-exception.filter';
import { ConflictResolutionService } from '../conflict/conflict.service';
import { DevicesRepository } from '../devices/devices.repository';
import type { OrgMemberContext } from '../organizations/org-context.types';
import { PermissionsService } from '../permissions/permissions.service';
import { CONNECTION_WRITABLE_FIELDS, CreateConnectionDto, PatchConnectionDto } from './connections.dto';
import { ConnectionsRepository } from './connections.repository';

@Injectable()
export class ConnectionsService {
  constructor(
    private readonly connectionsRepository: ConnectionsRepository,
    private readonly devicesRepository: DevicesRepository,
    private readonly conflictService: ConflictResolutionService,
    private readonly audit: AuditService,
    private readonly permissions: PermissionsService,
  ) {}

  async listConnections(member: OrgMemberContext, deviceId?: string): Promise<PaginatedResponse<DeviceConnectionDto>> {
    // `deviceId` is a raw query param; a malformed value (e.g. the array Express
    // parses from `?deviceId[]=a&deviceId[]=b`) would otherwise reach Prisma as an
    // invalid scalar filter and 500. A bad filter is client input → 400.
    if (deviceId !== undefined && !isUUID(deviceId)) {
      throw new NodeScopeException('GEN_001', 'INVALID_DEVICE_ID', HttpStatus.BAD_REQUEST);
    }
    const scope = await this.permissions.scopeFilter(member);
    const items = await this.connectionsRepository.listVisible(member.organizationId, scope, deviceId);
    return { items: items.map((c) => this.toDto(c)), total: items.length };
  }

  async createConnection(
    member: OrgMemberContext,
    creatorUserId: string,
    dto: CreateConnectionDto,
  ): Promise<DeviceConnectionDto> {
    const organizationId = member.organizationId;

    if (dto.sourceDeviceId === dto.targetDeviceId) {
      throw new NodeScopeException('CONN_002', 'SELF_CONNECTION', HttpStatus.UNPROCESSABLE_ENTITY);
    }

    const { sourceSiteId, targetSiteId } = await this.resolveEndpointSites(
      organizationId,
      dto.sourceDeviceId,
      dto.targetDeviceId,
    );
    await this.assertBothEndpoints(member, sourceSiteId, targetSiteId);

    const duplicate = await this.connectionsRepository.existsDuplicate(
      organizationId,
      dto.sourceDeviceId,
      dto.targetDeviceId,
      dto.connectionType,
    );
    if (duplicate) {
      throw new NodeScopeException('CONN_003', 'DUPLICATE_CONNECTION', HttpStatus.CONFLICT);
    }

    const connection = await this.connectionsRepository.create({
      organizationId,
      userId: creatorUserId,
      ...dto,
    });
    await this.audit.recordCreate(organizationId, 'DeviceConnection', connection);
    return this.toDto(connection);
  }

  async getConnection(member: OrgMemberContext, connectionId: string): Promise<DeviceConnectionDto> {
    const scope = await this.permissions.scopeFilter(member);
    const connection = await this.connectionsRepository.findVisibleByIdAndOrgId(connectionId, member.organizationId, scope);
    if (!connection) {
      throw new NodeScopeException('CONN_001', 'CONNECTION_NOT_FOUND', HttpStatus.NOT_FOUND);
    }
    return this.toDto(connection);
  }

  async updateConnection(
    member: OrgMemberContext,
    connectionId: string,
    patch: PatchConnectionDto,
  ): Promise<DeviceConnectionDto> {
    const organizationId = member.organizationId;
    // Write path: org-wide lookup so out-of-scope ADMIN → PERM_001 not 404
    const connection = await this.connectionsRepository.findByIdAndOrgId(connectionId, organizationId);
    if (!connection) {
      throw new NodeScopeException('CONN_001', 'CONNECTION_NOT_FOUND', HttpStatus.NOT_FOUND);
    }

    const { sourceSiteId, targetSiteId } = await this.resolveEndpointSites(
      organizationId,
      connection.sourceDeviceId,
      connection.targetDeviceId,
    );
    await this.assertBothEndpoints(member, sourceSiteId, targetSiteId);

    const updatePayload = this.conflictService.buildUpdatePayload(
      patch,
      CONNECTION_WRITABLE_FIELDS,
      connection.version,
      CreateConnectionDto,
    );

    const updated = await this.connectionsRepository.updateWithVersion(
      connectionId,
      organizationId,
      updatePayload,
      patch.baseVersion,
    );
    if (!updated) {
      throw new NodeScopeException('SYNC_001', 'EDIT_CONFLICT', HttpStatus.CONFLICT);
    }

    const dto = this.toDto(updated);
    await this.audit.recordUpdate(organizationId, 'DeviceConnection', connectionId, patch.changes);
    this.conflictService.emitEntityEvent(
      WS_EVENTS.CONNECTION_UPDATED,
      { connectionId, connection: dto, changes: patch.changes, updatedBy: updated.userId ?? '' },
      organizationId,
    );
    return dto;
  }

  async deleteConnection(member: OrgMemberContext, connectionId: string): Promise<void> {
    const organizationId = member.organizationId;
    // Write path: org-wide lookup so out-of-scope ADMIN → PERM_001 not 404
    const connection = await this.connectionsRepository.findByIdAndOrgId(connectionId, organizationId);
    if (!connection) {
      throw new NodeScopeException('CONN_001', 'CONNECTION_NOT_FOUND', HttpStatus.NOT_FOUND);
    }

    const { sourceSiteId, targetSiteId } = await this.resolveEndpointSites(
      organizationId,
      connection.sourceDeviceId,
      connection.targetDeviceId,
    );
    await this.assertBothEndpoints(member, sourceSiteId, targetSiteId);

    await this.connectionsRepository.deleteByIdAndOrgId(connectionId, organizationId);
    await this.audit.recordDelete(organizationId, 'DeviceConnection', connection);
    this.conflictService.emitEntityEvent(
      WS_EVENTS.CONNECTION_DELETED,
      { connectionId },
      organizationId,
    );
  }

  /** Load both endpoint devices (throws DEVICE_001 if either missing) and return their siteIds. */
  private async resolveEndpointSites(
    organizationId: string,
    sourceDeviceId: string,
    targetDeviceId: string,
  ): Promise<{ sourceSiteId: string; targetSiteId: string }> {
    const [source, target] = await Promise.all([
      this.devicesRepository.findByIdAndOrgId(sourceDeviceId, organizationId),
      this.devicesRepository.findByIdAndOrgId(targetDeviceId, organizationId),
    ]);
    if (!source || !target) {
      throw new NodeScopeException('DEVICE_001', 'DEVICE_NOT_FOUND', HttpStatus.NOT_FOUND);
    }
    return { sourceSiteId: (source as Device).propertyId, targetSiteId: (target as Device).propertyId };
  }

  /** Assert write access to both endpoint sites. MEMBER → ORG_003; ADMIN missing either → PERM_001. */
  private async assertBothEndpoints(
    member: OrgMemberContext,
    sourceSiteId: string,
    targetSiteId: string,
  ): Promise<void> {
    await this.permissions.assertCanConfigure(member, sourceSiteId);
    await this.permissions.assertCanConfigure(member, targetSiteId);
  }

  private toDto(connection: DeviceConnection): DeviceConnectionDto {
    return {
      id: connection.id,
      userId: connection.userId,
      sourceDeviceId: connection.sourceDeviceId,
      targetDeviceId: connection.targetDeviceId,
      connectionType: connection.connectionType as DeviceConnectionDto['connectionType'],
      notes: connection.notes,
      version: connection.version,
      createdAt: connection.createdAt.toISOString(),
      updatedAt: connection.updatedAt.toISOString(),
    };
  }
}

File: src/Modules/Inventory/NodeScope.Modules.Inventory.Application/Links/ConnectionsService.cs
using NodeScope.Contracts.Realtime;
using NodeScope.Modules.Inventory.Domain;
using NodeScope.Platform.Abstractions;

namespace NodeScope.Modules.Inventory.Application.Links;

/// <summary>Device-connection CRUD, ported from Node's <c>ConnectionsService</c>.</summary>
public sealed class ConnectionsService
{
    private readonly IConnectionRepository _connections;
    private readonly LinkEndpoints _endpoints;
    private readonly IPermissionScopeService _permissions;
    private readonly IRealtimeService _realtime;
    private readonly IAuditService _audit;

    public ConnectionsService(
        IConnectionRepository connections,
        LinkEndpoints endpoints,
        IPermissionScopeService permissions,
        IRealtimeService realtime,
        IAuditService audit)
    {
        _connections = connections;
        _endpoints = endpoints;
        _permissions = permissions;
        _realtime = realtime;
        _audit = audit;
    }

    public async Task<LinkListDto<ConnectionDto>> ListAsync(
        OrgMemberContext member,
        string? deviceId,
        CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(member);
        if (deviceId is not null && !Guid.TryParse(deviceId, out _))
        {
            throw new ApiException("GEN_001", "INVALID_DEVICE_ID", 400);
        }

        var scope = await _permissions.ScopePropertyIdsAsync(member, cancellationToken);
        var items = await _connections.ListVisibleAsync(member.OrganizationId, scope, deviceId, cancellationToken);
        return new LinkListDto<ConnectionDto>([.. items.Select(connection => connection.ToDto())], items.Count);
    }

    public async Task<ConnectionDto> CreateAsync(
        OrgMemberContext member,
        string creatorUserId,
        CreateConnectionRequest request,
        CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(member);
        ArgumentNullException.ThrowIfNull(request);
        if (request.SourceDeviceId == request.TargetDeviceId)
        {
            throw InventoryErrors.SelfConnection();
        }

        var organizationId = member.OrganizationId;
        await _endpoints.AuthorizeBothAsync(
            member, request.SourceDeviceId!, request.TargetDeviceId!, cancellationToken);

        var connectionType = ConnectionTypeLabels.TryParse(request.ConnectionType)!.Value;
        if (await _connections.DuplicateExistsAsync(
            organizationId, request.SourceDeviceId!, request.TargetDeviceId!, connectionType, cancellationToken))
        {
            throw InventoryErrors.DuplicateConnection();
        }

        var connection = await _connections.CreateAsync(
            new NewConnection(
                organizationId,
                creatorUserId,
                request.SourceDeviceId!,
                request.TargetDeviceId!,
                connectionType,
                request.Notes),
            cancellationToken);
        var dto = connection.ToDto();
        await _audit.RecordCreateAsync(organizationId, "DeviceConnection", connection.Id, dto, cancellationToken);
        return dto;
    }

    public async Task<ConnectionDto> GetAsync(
        OrgMemberContext member,
        string connectionId,
        CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(member);
        var scope = await _permissions.ScopePropertyIdsAsync(member, cancellationToken);
        var connection = await _connections.FindVisibleAsync(
            member.OrganizationId, connectionId, scope, cancellationToken)
            ?? throw InventoryErrors.ConnectionNotFound();
        return connection.ToDto();
    }

    public async Task<ConnectionDto> UpdateAsync(
        OrgMemberContext member,
        string connectionId,
        ChangesetRequest patch,
        CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(member);
        ArgumentNullException.ThrowIfNull(patch);
        var organizationId = member.OrganizationId;
        // Write path: org-wide lookup so an out-of-scope ADMIN gets PERM_001, not 404.
        var connection = await _connections.FindAsync(organizationId, connectionId, cancellationToken)
            ?? throw InventoryErrors.ConnectionNotFound();
        var sites = await _endpoints.AuthorizeBothAsync(
            member, connection.SourceDeviceId, connection.TargetDeviceId, cancellationToken);

        var payload = Changesets.BuildUpdatePayload(
            patch, ConnectionFields.Writable, connection.Version, ConnectionFields.IsValid);
        var updated = await _connections.UpdateWithVersionAsync(
            organizationId, connectionId, payload, patch.BaseVersion!.Value, cancellationToken)
            ?? throw Changesets.EditConflict();

        var dto = updated.ToDto();
        await _audit.RecordUpdateAsync(
            organizationId, "DeviceConnection", connectionId, ChangesetAudit.ToFieldChanges(patch), cancellationToken);
        await _realtime.EmitScopedMultiAsync(
            organizationId,
            [sites.First, sites.Second],
            WsEvents.ConnectionUpdated,
            new
            {
                connectionId,
                connection = dto,
                changes = patch.Changes,
                updatedBy = updated.UserId ?? "",
                timestamp = IsoTimestamp.Now(),
            },
            cancellationToken);
        return dto;
    }

    public async Task DeleteAsync(OrgMemberContext member, string connectionId, CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(member);
        var organizationId = member.OrganizationId;
        var connection = await _connections.FindAsync(organizationId, connectionId, cancellationToken)
            ?? throw InventoryErrors.ConnectionNotFound();
        var sites = await _endpoints.AuthorizeBothAsync(
            member, connection.SourceDeviceId, connection.TargetDeviceId, cancellationToken);

        await _connections.DeleteAsync(organizationId, connectionId, cancellationToken);
        await _audit.RecordDeleteAsync(
            organizationId, "DeviceConnection", connectionId, connection.ToDto(), cancellationToken);
        await _realtime.EmitScopedMultiAsync(
            organizationId,
            [sites.First, sites.Second],
            WsEvents.ConnectionDeleted,
            new { connectionId, timestamp = IsoTimestamp.Now() },
            cancellationToken);
    }
}
